WebRTC r1.1 - Release candidate preparations

stroncoso-quobis opened the pull request:
#### What type of PR is this?

Add one of the following kinds:
* documentation
* subproject management

#### What this PR does / why we need it:

* Release process to create first public release candidate for WebRTC APIs
* Target release r0.1.1 (not tagged yet)

#### Which issue(s) this PR fixes:

Fixes #50 
Fixes #51
Fixes #61 
Fixes #60 
Fixes #58 

#### Special notes for reviewers:

This PR intend to solve any formality and documentation requirements to achieve public release status within the repository and APIs: webrtc-registration, webrtc-call-handling, and webrtc-events. No functionality changes are intended.

#### Changelog input

```
Version URL path to rc.1
- webrtc-registration: 0.2.0-rc.1
- webrtc-call-handling: 0.2.0-rc.1
- webrtc-events: 0.1.0-rc.1
```

#### Additional documentation 

```
API readiness checklist for webrtc-registration
API readiness checklist for webrtc-call-handling
API readiness checklist for webrtc-events
```

hdamker commented:
@stroncoso-quobis Three points on first glance:

* The Readiness Checklist should reflect the status of the planned release ... currently the status column is empty
* A rc version of the APIs requires at least the basic test case definition, without them an alpha release is possible (see "mandatory" "Basic API test cases & documentation" in the Checklist). 
* For Spring25 the Commonalities version is 0.5, not 0.4.0

hdamker commented:
Regarding release numbering: For the Spring25 release the first (pre)release will be r1.1.

Regarding API versioning: a v0.1.3 would be only a patch release of the existing v0.1.1 ... are you sure that there were no breaking changes at all? (e.g. by applying the Commonality guidelines?)
One change would be for example that the URL for a CAMARA 0.x.y API should be .../v0.x, not /v0 (to be able to distinct it from a further 0.x+1.y version) also from client side.
My recommendation would be to go to 0.2.0 for the Spring25 release cycle.

rartych commented on code/API_definitions/webrtc-registration.yaml:
```diff
@@ -12,14 +12,16 @@ info:
     url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: 0.1.3-rc.1
+  x-camara-commonalities: 0.4.0
+  
```
Removing trailing spaces
```suggestion

```

rartych commented on documentation/API_documentation/webrtc-registration-API-Readiness-Checklist.md:
```diff
@@ -0,0 +1,18 @@
+|  2 | Design guidelines from Commonalities applied |   O   |   M   |   M   |   M   |  tbd  | 0.5   |
+|  3 | Guidelines from ICM applied                  |   O   |   M   |   M   |   M   |  tbd  | 0.3.0 |
+|  4 | API versioning convention applied            |   M   |   M   |   M   |   M   |   Y   | SemVer 2.0.0 |
+|  5 | API documentation                            |   M   |   M   |   M   |   M   |   Y   | [link](/documentation/API_documentation/webrtc-registration-API-Readiness-Checklist.md) |
```
The documentation should be inline in the YAML file `info:description` field, so it can be rendered with Redoc or Swagger like this:
https://editor.swagger.io/?url=https://raw.githubusercontent.com/camaraproject/WebRTC/refs/heads/main/code/API_definitions/webrtc-registration.yaml

The diagrams can be linked from Github repository (documentation folder)

As a result YAML should cover API specification and Documentation - see [QoD API](https://editor.swagger.io/?url=https://raw.githubusercontent.com/camaraproject/QualityOnDemand/r1.3/code/API_definitions/quality-on-demand.yaml)
```suggestion
|  5 | API documentation                            |   M   |   M   |   M   |   M   |   Y   | inline in YAML|
```
Initially it can be linked to `API_documentation` folder https://github.com/camaraproject/WebRTC/tree/main/documentation/API_documentation

stroncoso-quobis replied:
Cristal clear, I will try to fix it on the next commit. Thanks!

tanjadegroot commented:
Hi, I think you may want to do some dedicated PR for each of the issues you listed in the beginning of this issue, and then mergen them. this would also give you a list of inputs for the changelog.

tanjadegroot commented on code/API_definitions/webrtc-events.yaml:
```diff
@@ -169,14 +169,14 @@ info:
   url: https://github.com/camaraproject/
 servers:
-  - url: "{apiRoot}/webrtc-events/vwip"
+  - url: "{apiRoot}/webrtc-events/v0.1"
```
the url version should be: v0.1rc1

tanjadegroot commented on code/API_definitions/webrtc-registration.yaml:
```diff
@@ -12,14 +12,15 @@ info:
   license:
     name: Apache 2.0
     url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: 0.2-rc.1
```
the version should be: 0.2.0-rc.1

tanjadegroot commented on code/API_definitions/webrtc-registration.yaml:
```diff
@@ -12,14 +12,15 @@ info:
+  x-camara-commonalities: 0.4.0
 servers:
-  - url: '{apiRoot}/webrtc-registration/vwip'
+  - url: '{apiRoot}/webrtc-registration/v0.2'
```
the url version shoudl be v0.2rc1

tanjadegroot commented on code/API_definitions/webrtc-call-handling.yaml:
I assume similar comments as the above 2 would apply to the other API OAS files - I'll let you cross check this.

tanjadegroot reviewed (commented):
I have quite a lot of comments for making a more CAMARA-like API. I only covered one OAS file, but I assume these comment will also hold for the other APIs. 

Please discuss on how to go forward with this API.

stroncoso-quobis commented on code/API_definitions/webrtc-call-handling.yaml:
```diff
@@ -17,9 +17,11 @@ info:
     url: 'http://www.apache.org/licenses/LICENSE-2.0.html'
-  version: 0.1.3
+  version: 0.2-rc.1
+  x-camara-commonalities: 0.4.0
```
Thanks. It is a WIP item.

Check progress at https://github.com/camaraproject/WebRTC/pull/65 PR. 0.5 comms will be covered as soon as it is merged.

stroncoso-quobis commented on code/API_definitions/webrtc-call-handling.yaml:
Hi @tanjadegroot , 

Thanks for the review. Let's try to add some responses, and we will see how to fix them:

> _Replace WebRTCGW by network_

Sounds good to me, I will take a look over it. All documentation needs to be reviewed.

> _replace vvoipSessionId as "vvoip" is a Telco term_

`vvoipSessionId` and `racmSessionId` are used for different kind of "sessions", one is the media session itself, the other one is the active registration of the user/device on the network. I think that is better to keep some difference between them.

> _The term "webrtc" is also Telco stuff_

In my humble opinion, I disagree, "webrtc" is a developer or technology term. It comes from the [WebRTC API](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API) , and it is widely used on the real-time developers landscape, WHIP/WHEP is an example of protocol that uses that techonlogy and most of users are familiar with the SDP terminology. At least when dealing with "calls".

`call-regsitration`, `call-handling` and `call-events` can be possible names, but "webrtc" does not look to me as a "Telco stuf", is more a "CPaaS stuff".

> _Unless you want to evolve this API later to support more than 2 participants_

I will really look for this, and yes, trying to be inspired by the most popular CPaaS, in order to bring easy transition from private CPaaS to Telco network systems.

> _line 9: replace RACM with new name_

Fair enough :) RACM stands for **R**egistration **A**nd **C**onnectivity **M**anager, it could be simplier, like "registration service" or "webrtc registration", we will try to came up with something new.

> _the term "hdr" is not explained_
> _hdrClientId and 340 hdrTransactioId: are these really informations that need to be exposed to the applications ?_

These terms probably came from previous specs that weren't adapted to the current usage of `x-correlator` and `deviceId`. We will check them. For sure *hdr* means *header* and we will think on something to change it.

> _line 266: WrtcSDPDescriptor [...] you really want to avoid a developer having to read the SDP RFC ..._

Yes, we need to avoid the SDP direct manipulation, it must not be a requirement. But is needed to provide a protocol to share the media descriptor of each side of the Real-Time communication.

As said before, is my belief that SDP is really common nowadays for RTC developers. That I consider the target of the API. What we must include on the documentation, is to provide clear instructions about how to obtain this SDP. Maybe some reference to [RTCSessionDescription](https://developer.mozilla.org/en-US/docs/Web/API/RTCSessionDescription) or other mechanisms.

We will take a look to this feedback.

> _If this is the attribute in the API POST allows to choose VOICE or VIDEO for the call to be created, could you make that the more user-friendly ? e.g. with a CallType parameter with a schema with an ENUM ? Maybe have a look at the APIs created by some of the CPaaS players for their Voice and Video calling ?_

There are a lot of room for improvement, this audio/video request is one of this possible improvements, there's no specification for that on the API. You can respond with what you want, based on the SDP description. 

For future releases, we need to review what is being done out there and try to get improvements. But this basic usage, within the CAMARA initiative will ease the developer pain for all the different CPaaS and GWs on the market.

We will discuss your feedback and create issues for the possible changes agreed.
Thanks a lot for your feedback.

Regards,

tanjadegroot replied:
Hi @stroncoso-quobis? I am happy to stand corrected on my view that WebRTC is a Telco term. If this is what is being used the the CPaaS context, then it makes sense to adopt related terminology. Maybe this is also due the abbreviations :-)

However looking at terminology on your link (https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API#webrtc_reference),
I think it would still be important to find more "human" terms, e.g. inspired by some of the terms used on that page but without the "RTC" part. For example having 
- api name: "real-time-communications" 
- url: "{apiRoot}/real-time-communications/vx" 
- api: "/sessions"
- parameter: sessionDescription
- etc.

That would 
- make the link with WebRTC for those who know it (and documented in the description)
- avoid abbreviations in the APIs and parameters
- make it more easily readable
- is generic enough to allow evolution to more features later

you may then be able to reach yet more developers :-)

Up to the team to discuss for sure.

tanjadegroot replied:
Hi @stroncoso-quobis, thanks for the feedback.
The main requirements are the alignment to the Commonalities and ICM to have consistent APIs in CAMARA.
All the rest is up to the team to make the API as usable as possible and when you think it is appropriate.

tanjadegroot commented:
Suggestion from Release mgmt team: we consider that Spring25 meta-release is too early given the current state of the API. 
Prepare the release for Fall25.

stroncoso-quobis commented:
> Suggestion from Release mgmt team: we consider that Spring25 meta-release is too early given the current state of the API. Prepare the release for Fall25.

Hi @tanjadegroot 

Can we discuss this recommendation?
We are gaining speed:
- Today's meeting agreed to merge #65 , so commonalities will be covered for all APIs.
- Today was pushed extra documentation for registration API PR #67 , and tomorrow, documentation for call-handling API is expected. It was previously shared during the meeting and the team considered it ok.
- Only remaining point pending is testing documentation, and I consider that we can comply by the next week.

During [today's meeting](https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/76349516/2025-02-11+WebRTC+Minutes?focusedTaskId=6), deadlines were commented, and we strongly believe that we can still achieve the deadline.

Please, in name of all the team, reconsider the suggestion and give us some deadline to deliver doc & testing.

Best regards,

stroncoso-quobis commented on documentation/API_documentation/webrtc-call-handling-API-Readiness-Checklist.md:
```diff
@@ -0,0 +1,18 @@
+|  4 | API versioning convention applied            |   M   |   M   |   M   |   M   |   Y   | SemVer 2.0.0 |
+|  5 | API documentation                            |   M   |   M   |   M   |   M   |   Y   | [link](/documentation/API_documentation/webrtc-call-handling-API-Readiness-Checklist.md) |
+|  6 | User stories                                 |   O   |   O   |   O   |   M   |   N   |   |
+|  7 | Basic API test cases & documentation         |   O   |   M   |   M   |   M   |  tbd  | [link](/documentation/API_documentation/) |
```
Included at 90b5d9b8679df9395b65faf18a0e6b9180bd300b and 3084d77302b31d32538e7cd4983ffe372222a8fc at PR #69

stroncoso-quobis commented on documentation/API_documentation/webrtc-events-API-Readiness-Checklist.md:
```diff
@@ -0,0 +1,18 @@
+|  4 | API versioning convention applied            |   M   |   M   |   M   |   M   |   Y   | SemVer 2.0.0 |
+|  5 | API documentation                            |   M   |   M   |   M   |   M   |   Y   | [link](/documentation/API_documentation/webrtc-events-API-Readiness-Checklist.md) |
+|  6 | User stories                                 |   O   |   O   |   O   |   M   |   N   |   |
+|  7 | Basic API test cases & documentation         |   O   |   M   |   M   |   M   |  tbd  | [link](/documentation/API_documentation/) |
```
Included at 90b5d9b8679df9395b65faf18a0e6b9180bd300b and 5bba9bc6508bfd9641c28bb1da03b5a08ac517e3 at PR #69

stroncoso-quobis commented on documentation/API_documentation/webrtc-registration-API-Readiness-Checklist.md:
```diff
@@ -0,0 +1,18 @@
+|  4 | API versioning convention applied            |   M   |   M   |   M   |   M   |   Y   | SemVer 2.0.0 |
+|  5 | API documentation                            |   M   |   M   |   M   |   M   |   Y   | [link](/documentation/API_documentation/webrtc-registration-API-Readiness-Checklist.md) |
+|  6 | User stories                                 |   O   |   O   |   O   |   M   |   N   |   |
+|  7 | Basic API test cases & documentation         |   O   |   M   |   M   |   M   |  tbd  | [link](/documentation/API_documentation/) |
```
Inlcuded at 90b5d9b8679df9395b65faf18a0e6b9180bd300b and 527304aa238f0652603adadf1a3da8a803c980ae at PR #69

stroncoso-quobis commented:
Hi team,

cc. @tanjadegroot @hdamker 

Let me summarize the work during these days to try to speed up the release revision:

- Covered commonalities, from 0.4 to 0.5
  - https://github.com/camaraproject/WebRTC/pull/66
  - https://github.com/camaraproject/WebRTC/pull/65
- Fixed urls, API versioning and release versioning (r1.1)
  - https://github.com/camaraproject/WebRTC/commit/cbb1e681a3cf15c095e6961bfaab9e5c9b501724
- Included YAML inline documentation for all APIs 
  - https://github.com/camaraproject/WebRTC/pull/67
  - https://github.com/camaraproject/WebRTC/pull/68
- Resolved branch conflicts.
  - https://github.com/camaraproject/WebRTC/commit/b7619103d4f65def5adc2c2de55f2d1deaf26fb5
- Included updated CHANGELOG and README on the release PR 
  - https://github.com/camaraproject/WebRTC/commit/be573aa80772ad5beb183592d519067a0283e7d2

Reagarding Wiki assets:
- Updated API trackers
  - https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/14564612/WebRTC+API+Release+Tracking
  - I don't known why, but sumary page does not include the pre-release tag included on the individual API trackers

Reviewing note:
- Since https://github.com/camaraproject/WebRTC/pull/62 is not merged and release not officially created, I just tag `r1.1` to main, in order to confirm that all links, references and materials point to the desired status of the repo.

Please, let me known any further action.

Best regards,

hdamker commented:
> * Updated API trackers
>   
>   * https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/14564612/WebRTC+API+Release+Tracking
>   * I don't known why, but sumary page does not include the pre-release tag included on the individual API trackers
> 
The table on the summary page wasn't updated to the new release tracker structure. I have corrected that.
But: it is too early to publish already the release link, as this is currently actually does not have the correct release content (which will come with this PR here). And the link to be used will be https://github.com/camaraproject/WebRTC/releases/tag/r1.1 (pointing to the release description first, not directly to the tag).

> Reviewing note:
> 
> * Since [WebRTC r1.1 - Release candidate preparations #62](https://github.com/camaraproject/WebRTC/pull/62) is not merged and release not officially created, I just tag `r1.1` to main, in order to confirm that all links, references and materials point to the desired status of the repo.

Makes sense, but don't forget to delete the tag before you create the release (and don't publish this tag already ... I'm using for the same purpose a branch ... but also that has to be deleted before creating the release (tag)).

stroncoso-quobis commented:
Hi @hdamker 

Release ready: https://github.com/camaraproject/WebRTC/releases/tag/r1.1

API tracker updated with M3 date (202502-17):
* [WebRTC registration v0.2.0](https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/39354592/webrtc-registration+v0.2.0)
* [WebRTC call handling v0.2.0](https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/39354632/webrtc-call-handling+v0.2.0)
* [WebRTC events v0.1.0](https://lf-camaraproject.atlassian.net/wiki/spaces/CAM/pages/39321972/webrtc-events+v0.1.0)

Best regards,

Thanks all for the efforts!!
